File: androidApp/src/main/java/io/lb/middleware/android/createroute/presentation/review/ReviewScreen.kt
package io.lb.middleware.android.createroute.presentation.review

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavHostController
import io.lb.middleware.android.core.presentation.Screens
import io.lb.middleware.android.core.presentation.components.GenericTopAppBar
import io.lb.middleware.android.core.presentation.components.MethodBox
import io.lb.middleware.android.core.presentation.components.RouteDetails
import io.lb.middleware.android.core.presentation.components.TestColumn
import io.lb.middleware.android.core.presentation.showToast
import io.lb.middleware.android.createroute.presentation.model.CreateRouteArgs
import io.lb.middleware.common.shared.middleware.request.MiddlewareHttpMethods
import io.lb.middleware.common.state.CommonFlow
import io.lb.middleware.shared.presentation.createroute.review.ReviewEvent
import io.lb.middleware.shared.presentation.createroute.review.ReviewState
import io.lb.middleware.shared.presentation.createroute.review.ReviewViewModel
import kotlinx.coroutines.flow.collectLatest

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReviewScreen(
    navController: NavHostController,
    args: CreateRouteArgs?,
    state: ReviewState,
    eventFlow: CommonFlow<ReviewViewModel.UiEvent>,
    onEvent: (ReviewEvent) -> Unit
) {
    val context = LocalContext.current
    val result = remember {
        mutableStateOf("")
    }
    val isFinished = remember {
        mutableStateOf(false)
    }

    LaunchedEffect(key1 = Screens.REVIEW) {
        eventFlow.collectLatest {
            when (it) {
                is ReviewViewModel.UiEvent.ShowCreatedRouteMessage -> {
                    context.showToast("Route Created!")
                    isFinished.value = true
                    result.value = it.result
                }

                is ReviewViewModel.UiEvent.ShowError -> {
                    context.showToast(it.message)
                }
            }
        }
    }

    Scaffold(
        modifier = Modifier.fillMaxSize(),
        topBar = {
            GenericTopAppBar(navController, "Step 5/5: Review & Create Route")
        },
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .padding(padding)
                .padding(horizontal = 16.dp)
                .fillMaxSize(),
            verticalArrangement = Arrangement.Top,
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            item {
                Column(
                    verticalArrangement = Arrangement.Top,
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {
                    Text(
                        modifier = Modifier.fillMaxWidth(),
                        text = "Original Base URL",
                        fontSize = 16.sp,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Spacer(modifier = Modifier.padding(4.dp))
                    Text(
                        modifier = Modifier.fillMaxWidth(),
                        text = args?.originalBaseUrl?.substringAfterLast("://") ?: "",
                        fontSize = 16.sp,
                    )
                    Spacer(modifier = Modifier.padding(12.dp))
                }
            }

            item {
                Column(
                    verticalArrangement = Arrangement.Top,
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {
                    Text(
                        modifier = Modifier.fillMaxWidth(),
                        text = "Original Path",
                        fontSize = 16.sp,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Spacer(modifier = Modifier.padding(4.dp))
                    Text(
                        modifier = Modifier.fillMaxWidth(),
                        text = args?.originalPath?.substringAfterLast("://") ?: "",
                        fontSize = 16.sp,
                    )
                    Spacer(modifier = Modifier.padding(8.dp))
                }
            }

            item {
                Column(
                    verticalArrangement = Arrangement.Top,
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {
                    Text(
                        modifier = Modifier.fillMaxWidth(),
                        text = "Mapped Path",
                        fontSize = 16.sp,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Text(
                        modifier = Modifier.fillMaxWidth(),
                        text = args?.mappedPath?.substringAfterLast("://") ?: "",
                        fontSize = 16.sp,
                        overflow = TextOverflow.Ellipsis,
                    )
                    Spacer(modifier = Modifier.padding(12.dp))
                }
            }

            item {
                Column(
                    verticalArrangement = Arrangement.Top,
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        MethodBox(
                            args?.originalMethod,
                            "Original Methods",
                            Modifier.width(80.dp)
                        )
                        Spacer(modifier = Modifier.padding(16.dp))
                        MethodBox(
                            args?.originalMethod,
                            "Mapped Methods",
                            Modifier.width(80.dp)
                        )
                    }
                    Spacer(modifier = Modifier.padding(12.dp))
                }
            }

            item {
                RouteDetails(
                    originalQueries = args?.originalQueries,
                    originalHeaders = args?.originalHeaders,
                    originalBody = args?.originalBody,
                    preConfiguredQueries = args?.preConfiguredQueries,
                    preConfiguredHeaders = args?.preConfiguredHeaders,
                    preConfiguredBody = args?.preConfiguredBody,
                )
            }

            item {
                TestColumn(
                    isLoading = state.isLoading && isFinished.value.not(),
                    result = result.value,
                    idleText = "Create Route",
                    progressText = "Creating Route",
                ) {
                    result.value = ""
                    onEvent(
                        ReviewEvent.CreateMappedRoute(
                            originalBaseUrl = args?.originalBaseUrl ?: "",
                            originalPath = args?.originalPath ?: "",
                            originalMethod = args?.originalMethod ?: MiddlewareHttpMethods.Get,
                            originalBody = args?.originalBody ?: "",
                            originalQueries = args?.originalQueries ?: emptyMap(),
                            originalHeaders = args?.originalHeaders ?: emptyMap(),
                            newBodyFields = args?.newBodyFields?.mapValues {
                                it.value.toNewBodyField()
                            } ?: emptyMap(),
                            oldBodyFields = args?.oldBodyFields?.mapValues {
                                it.value.toOldBodyField()
                            } ?: emptyMap(),
                            preConfiguredQueries = args?.preConfiguredQueries ?: emptyMap(),
                            preConfiguredHeaders = args?.preConfiguredHeaders ?: emptyMap(),
                            mappedPath = args?.mappedPath ?: "",
                            mappedMethod = args?.mappedMethod ?: MiddlewareHttpMethods.Get,
                            preConfiguredBody = args?.preConfiguredBody ?: "",
                            ignoreEmptyValues = args?.ignoreEmptyFields ?: true
                        )
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/create_route/review/ReviewScreen.swift
//
//  ReviewScreen.swift
//  iosApp
//
//  Created by Leonardo Bai on 06/04/25.
//  Copyright © 2025 orgName. All rights reserved.
//

import SwiftUI
import Shared

struct ReviewScreen: View {
    private var args: CreateRouteArgs
    private var createNewRouteUseCase: Middleware_domainCreateNewRouteUseCase
    @ObservedObject private var viewModel: IOSReviewViewModel
    
    @State private var result = ""
    @State private var isFinished = false
    @State private var showErrorAlert = false
    @State private var errorMessage = ""
    @State private var eventSubscription: StateDisposableHandle?
    @Environment(\.navigate) private var navigate
    @Environment(\.replace) private var replace
    
    init(
        args: CreateRouteArgs,
        createNewRouteUseCase: Middleware_domainCreateNewRouteUseCase
    ) {
        self.args = args
        self.createNewRouteUseCase = createNewRouteUseCase
        self.viewModel = IOSReviewViewModel(
            createNewRouteUseCase: createNewRouteUseCase
        )
    }
    
    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                originalBaseUrlSection
                originalPathSection
                mappedPathSection
                methodsSection
                routeDetailsSection
                testSection
            }
            .padding(.vertical)
        }
        .onAppear { setupScreen() }
        .onDisappear { cleanup() }
        .background(Color.background)
        .toolbar {
            if isFinished {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button(
                        action: {
                            replace(.routeListing)
                        }
                    ) {
                        Text(isFinished ? "Done" : "Back")
                            .foregroundColor(Color.primaryColor)
                    }
                }
            }
        }
        .alert("Error", isPresented: $showErrorAlert) {
            Button("OK", role: .cancel) { }
        } message: {
            Text(errorMessage)
        }
        .navigationTitle("Step 5/5: Review & Create Route")
        .navigationBarBackButtonHidden(isFinished)
    }
    
    // MARK: - Subviews
    
    private var originalBaseUrlSection: some View {
        VStack(alignment: .leading) {
            Text("Original Base URL")
                .font(.system(size: 16))
                .foregroundColor(Color.primaryColor)
            
            Text(args.originalBaseUrl.components(separatedBy: "://").last ?? "")
                .font(.system(size: 16))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal)
    }
    
    private var originalPathSection: some View {
        VStack(alignment: .leading) {
            Text("Original Path")
                .font(.system(size: 16))
                .foregroundColor(Color.primaryColor)
            
            Text(args.originalPath.components(separatedBy: "://").last ?? "")
                .font(.system(size: 16))
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal)
    }
    
    private var mappedPathSection: some View {
        VStack(alignment: .leading) {
            Text("Mapped Path")
                .font(.system(size: 16))
                .foregroundColor(Color.primaryColor)
            
            Text(args.mappedPath.components(separatedBy: "://").last ?? "")
                .font(.system(size: 16))
                .lineLimit(1)
                .truncationMode(.tail)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal)
    }
    
    private var methodsSection: some View {
        HStack(spacing: 24) {
            MethodBox(
                method: args.originalMethod,
                text: "Original Method"
            )
            MethodBox(
                method: args.mappedMethod,
                text: "Mapped Method"
            )
            Spacer()
        }
        .padding(.horizontal)
    }
    
    private var routeDetailsSection: some View {
        RouteDetails(
            originalQueries: args.originalQueries,
            originalHeaders: args.originalHeaders,
            originalBody: args.originalBody,
            preConfiguredQueries: args.preConfiguredQueries,
            preConfiguredHeaders: args.preConfiguredHeaders,
            preConfiguredBody: args.preConfiguredBody
        )
    }
    
    private var testSection: some View {
        TestColumn(
            isLoading: viewModel.state.isLoading,
            result: result,
            isFinished: isFinished,
            idleText: "Create Route",
            progressText: "Creating Route",
            onClick: {
                result = ""
                viewModel.onEvent(event: ReviewEvent.CreateMappedRoute(
                    originalBaseUrl: args.originalBaseUrl,
                    originalPath: args.originalPath,
                    originalMethod: args.originalMethod,
                    originalBody: args.originalBody,
                    originalQueries: args.originalQueries,
                    originalHeaders: args.originalHeaders,
                    newBodyFields: args.newBodyFields.mapValues { $0.toModel() },
                    oldBodyFields: args.oldBodyFields.mapValues { $0.toModel() },
                    preConfiguredQueries: args.preConfiguredQueries,
                    preConfiguredHeaders: args.preConfiguredHeaders,
                    mappedPath: args.mappedPath,
                    mappedMethod: args.mappedMethod,
                    preConfiguredBody: args.preConfiguredBody,
                    ignoreEmptyValues: args.ignoreEmptyFields
                ))
            }
        )
        .padding(.horizontal)
    }
    
    private func setupScreen() {
        viewModel.startObserving()
        subscribeToEvents()
    }
    
    private func cleanup() {
        viewModel.dispose()
        eventSubscription?.dispose()
    }
    
    private func subscribeToEvents() {
        eventSubscription = viewModel.eventFlow.subscribe(
            onCollect: { event in
                DispatchQueue.main.async {
                    guard let event = event else { return }
                    handleUiEvent(event)
                }
            }
        )
    }
    
    private func handleUiEvent(_ event: ReviewViewModel.UiEvent) {
        switch event {
        case let showError as ReviewViewModel.UiEventShowError:
            errorMessage = showError.message
            showErrorAlert = true
            
        case let showResult as ReviewViewModel.UiEventShowCreatedRouteMessage:
            result = showResult.result
            isFinished = true
            
        default:
            break
        }
    }
}
